daliuge-engine/dlg/apps/pyfunc.py

```diff
@@ -712,6 +712,19 @@ def write_results(self, result):
                 elif self.output_parser is DropParser.NPY:
                     drop_loaders.save_npy(o, r)
                 elif self.output_parser is DropParser.RAW:
+                    if not isinstance(r, (bytes, memoryview, str)):
+                        raise Exception(
+                            "Can't write data of this type: ", type(r).__name__
+                        )
+                    o.write(r)
+                elif self.output_parser is DropParser.UTF8:
+                    if not isinstance(r, (bytes, memoryview, str)):
+                        try:
+                            r = str(r)
+                        except:  # not sure what the exception would be...
+                            raise Exception(
+                                "Can't write data of this type: ", type(r).__name__
```

Comment:
This exception message is not technically correct - this exception is being raised because we can't convert `r` to type string.

Reply:
No, we don't want to convert to string here, we just want to make sure that we can write what was supplied on that port and is has to be one of these three types, without conversion, that is the whole point of 'RAW'.

Reply:
If that is the case, doesn't that make the `r = str(r)` conversion redundant? Either we want to raise the error because it isn't one of the three types it has to be, or we are happy converting it to str.